pkg/spanconfig/spanconfig.go

```diff
@@ -77,6 +78,47 @@ func FullTranslate(
 	return s.Translate(ctx, descpb.IDs{keys.RootNamespaceID})
 }
 
+// SQLWatcher watches for events on system.zones and system.descriptors.
+type SQLWatcher interface {
+	// WatchForSQLUpdates watches for changes to zones and descriptors starting at
+	// the given timestamp (exclusive) by establishing rangefeeds over
+	// system.zones and system.descriptors.
+	//
+	// It periodically calls the handle function with a list of SQLWatcherUpdates
+	// and a checkpointTS. Invocations of handle are guaranteed to provide the
+	// following semantics:
+	// 1. Calls to handle are serial.
+	// 2. The timestamp supplied to handle is monotonically increasing.
+	// 3. The list of SQLWatcherUpdates supplied to handle includes all events
+	// in the window (prevInvocationCheckpointTS, checkpointTS].
+	// 4. No further calls to handle are made if an invocation returns an error.
+	//
+	// These guarantees allow users of this interface to safely persist the
+	// checkpointTS and use it as the startTS for WatchForSQLUpdates later without
+	// missing any updates.
+	//
+	// All rangefeeds established by the SQLWatcher will be torn down before the
+	// call to WatchForSQLUpdates returns.
+	WatchForSQLUpdates(ctx context.Context, startTS hlc.Timestamp, handle SQLWatcherHandleFunc) error
+}
+
+// SQLWatcherHandleFunc is the type of the handler function periodically invoked
+// by the SQLWatcher.
+type SQLWatcherHandleFunc func(
+	ctx context.Context, updates []SQLWatcherUpdate, checkpointTS hlc.Timestamp,
+) error
+
+// SQLWatcherUpdate captures the ID and type of a descriptor or zone that the
+// SQLWatcher has observed change.
+type SQLWatcherUpdate struct {
+	// ID of the descriptor/zone that has changed.
+	ID descpb.ID
+
+	// DescriptorType of the descriptor/zone that has changed. Could be either the
+	// specific type or catalog.Any.
+	DescriptorType catalog.DescriptorType
```

Comment:
should this be a bitmask so that the events corresponding to a single ID can be combined?